---
title: "00_2D_gast_preprocess"
author: "achemkhi"
date: "2024-12-06"
output:
  html_document:
    toc: true
    toc_float: true
  pdf_document: default
    
bibliography: references.bib
---

All upstream analysis steps before running MIIC

-   data type : flash-seq

### Set version

```{r}
# this is version used to generate all the results and intermidiate files.
# please commit your changes after fixing the version and passing the results to MIIC 

P_VERSION="v0.2" # using GAEL seurat object 

OUTDIR=paste("~/Desktop/myProjects/2D_gast/output",P_VERSION,sep="/")


```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Seurat)
library(dplyr)
library(grid)
library(gridExtra)
library(here)
library(fs)
library(millefy)
setwd("~/Desktop/myProjects/2D_gast")
fs::link_create(path = "/Users/alichemkhi/Desktop/data", new_path = "data")# create symbolic link to data="/Users/alichemkhi/Desktop/data"
fs::link_create(path = "/Users/alichemkhi/Desktop/myProjects/renv", new_path = "renv")# create symbolic link to data="/Users/alichemkhi/Desktop/data"


```

## Load raw data

```{r}

# read count matrices
data.fs.read.1 <- as.matrix(Read10X(data.dir ="../data/2dgas-fs/raw_data/Plate1/analysis_results/Matrices/10XlikeMatrix_read_unzipped"))
data.fs.read.2.3 <- as.matrix(Read10X(data.dir ="../data/2dgas-fs/raw_data/Plate2-3/analysis_results/Matrices/10XlikeMatrix_read_unzipped"))

list.genes <- intersect(row.names(data.fs.read.1),row.names(data.fs.read.2.3))
data.fs.read <- cbind(data.fs.read.1[list.genes,],data.fs.read.2.3[list.genes,])

colnames(data.fs.read) <- substring(colnames(data.fs.read), first = 1, last = regexpr("_",unique(colnames(data.fs.read)))[1]-1)

data.fs.read.save <- data.fs.read
row.names(data.fs.read.save) <- toupper(row.names(data.fs.read.save))
write.table(data.fs.read.save,paste(OUTDIR,"RawReadMatrix.txt",sep ="/"))


```

# Transcript Count estimation

The total raw count is not equivalent to how many times the gene was transcribed because in flash-seq and full-length methods in general, you may have multiple reads coming from the same transcript (molecule) , we have to estimate the real number of transcripts .

## 1- Estimate transcript counts : Divide with length

```{r}
require(org.Mm.eg.db)

# In rsem.table gene length = sum(exons_length) ? 
rsem.table <- read.table('../data/2dgas-fs/raw_data/D1439T380_Sorted.genes.txt',header = T) # + length 
genes_id <- rsem.table$gene_id
genes_name <- mapIds(org.Mm.eg.db,
                     keys = genes_id,
                     column = 'SYMBOL',
                     keytype = 'ENSEMBL')
genes_name.table <- data.frame(row.names = NULL,
                  gene_id =names(genes_name),
                  SYMBOL  = genes_name)

# define length vector 
rsem.table.symbol <- left_join(rsem.table, genes_name.table, by = "gene_id")
genes_length=rsem.table.symbol$length
names(genes_length)<-rsem.table.symbol$SYMBOL
head(genes_length)
length(genes_length)
sum(is.na(genes_length)); 

# Add missing genes - not found in D1439T380_Sorted.genes.txt'
path_gtf = c("../data/2dgas-fs/raw_data/mm10_genes.gtf") 
dt_gtf_exon <- gtfToDtExon(path_gtf) # retrieving only exons 
dt_gtf_exon$gene_name <- substring(dt_gtf_exon$transcript_name,first=1, last = regexpr("-2",dt_gtf_exon$transcript_name)-1)

other.genes <- unique(dt_gtf_exon$gene_name)[!(unique(dt_gtf_exon$gene_name) %in% names(genes_length))]
for (gene.i in other.genes){
  gene.i.exon <- dt_gtf_exon[dt_gtf_exon$transcript_name==paste(gene.i,"-201",sep=""),] #NOTE :  get only the first canonical or primary isoform -201 to calculate the gene length 
  genes_length[gene.i] <- sum(gene.i.exon$end) - sum(gene.i.exon$start)
}
head(genes_length)
length(genes_length)
sum(is.na(genes_length))

# Prepare new matrix to populate with the normalized values 
data.fs.transcript <- matrix(NA, nrow = nrow(data.fs.read), ncol = ncol(data.fs.read))
colnames(data.fs.transcript) <- colnames(data.fs.read)
row.names(data.fs.transcript) <- row.names(data.fs.read)

all.genes=row.names(data.fs.transcript) # all genes in the count matrix  
genes.with.no.length<-all.genes[!all.genes %in% names(genes_length)]  # genes that don't have a length for the normalization step 
print("genes with no length info "); length(genes.with.no.length) 

# NORMALIZE w.r.t length 
for (gene.i in row.names(data.fs.transcript)) {
  data.fs.transcript[gene.i,] <- 100*data.fs.read[gene.i,]/genes_length[gene.i]
}
dim(data.fs.transcript)
data.fs.transcript <- na.omit(round(data.fs.transcript+0.33)) # remove na 
dim(data.fs.transcript)
write.table(data.fs.transcript,paste(OUTDIR,"TranscriptEstimationMatrix.txt",sep ="/"))

```

# UMI counts versus Reads counts

```{r}

data.fs.transcript <- as.matrix(read.table(paste(OUTDIR,"TranscriptEstimationMatrix.txt",sep ="/")))

#data.fs.transcript <- as.matrix(read.table(file = "~/Desktop/myProjects/2D_gast/data/2dgas-fs/raw_data/TranscriptEstimationMatrix.txt"))

# load umi counts 
# umi count matrices are expeted to be less than read count because it depends on the gene length , and the read including the UMI generally captures a tiny fraction of gene 
data.fs.umi.1 <- as.matrix(Read10X(data.dir ="../data/2dgas-fs/raw_data/Plate1/analysis_results/Matrices/10XlikeMatrix_umi_unzipped"))
data.fs.umi.2.3 <- as.matrix(Read10X(data.dir ="../data/2dgas-fs/raw_data/Plate2-3/analysis_results/Matrices/10XlikeMatrix_umi_unzipped"))
list.genes <- intersect(row.names(data.fs.transcript),intersect(row.names(data.fs.umi.1),row.names(data.fs.umi.2.3)))
data.fs.umi <- cbind(data.fs.umi.1[list.genes,],data.fs.umi.2.3[list.genes,])
data.fs.transcript.test <- data.fs.transcript[list.genes,]

data.x1 <- log1p(colSums(data.fs.transcript.test))
data.y1 <- log1p(colSums(data.fs.umi))
dat1 <- data.frame(x = data.x1, y = data.y1)

library(ggplot2)
library(stringr) 

size=2

# Transcript mean expression (log) versus y="UMI mean expression (log)
p <- ggplot(data = dat1, mapping = aes(x = x, y = y)) +
  geom_point()+
  theme_bw()+
  labs(x="Transcript mean expression (log)",y="UMI mean expression (log)")+
  theme(axis.text = element_text(size=size*5),
        axis.title = element_text(size=size*5),
        legend.key.size = unit(0.25*size, 'cm'),
        legend.title= element_text(size=size*5),
        legend.text = element_text(size=size*5),
        legend.title.align=0.5,
        panel.grid.minor = element_blank(),
        axis.line = element_line(colour = "black"))
p


```

# Create transcript seurat object

## Samples info

```{r}
data.fs.transcript <- as.matrix(read.table(paste(OUTDIR,"TranscriptEstimationMatrix.txt",sep ="/")))

sample.information <- read.table(file="../data/2dgas-fs/raw_data/sample_information.txt"
                                 ,header = TRUE)[colnames(data.fs.transcript),]

cells <- paste(sample.information$time,sample.information$condition,sep="-")
row.names(data.fs.transcript) <- toupper(row.names(data.fs.transcript))
colnames(data.fs.transcript) <- paste(cells,"_",colnames(data.fs.transcript ),sep="")

dataset.fs <- CreateSeuratObject(data.fs.transcript , project = "fs", min.cells = 1)
dataset.fs@meta.data$cells <- dataset.fs@active.ident
dataset.fs@meta.data$time <- as.factor(sample.information$time)
dataset.fs@meta.data$condition <- as.factor(sample.information$condition)

# value counts 
dataset.fs
table(dataset.fs@meta.data$orig.ident)
head(dataset.fs@meta.data)



```

## QC filtering

```{r}

s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

## QC and selecting cells for further analysis
# Get the percentage of mitochondrial genes in each cell
mito.genes <- grep(pattern = "^MT-", x = rownames(x = GetAssayData(dataset.fs, slot = "counts")), value = TRUE)
percent.mito <- Matrix::colSums(GetAssayData(dataset.fs, slot = "counts")[mito.genes, ])/Matrix::colSums(GetAssayData(dataset.fs, slot = "counts"))
dataset.fs <- AddMetaData(object = dataset.fs, metadata = percent.mito, col.name = "percent.mito")


VlnPlot(object = dataset.fs, features=c("nCount_RNA","nFeature_RNA","percent.mito"),slot = "counts",idents=c("t35-WNT", "t35-BMP","t36-WNT", "t36-BMP","t37-WNT", "t37-BMP"))

FeatureScatter(object = dataset.fs, feature1 = "nCount_RNA", feature2 = "nFeature_RNA",slot = "counts")+geom_vline(xintercept = 250000, color = "red", size=1)


# NOTE : filter 
dataset.fs <- subset(dataset.fs, subset =  nFeature_RNA > 200 & nCount_RNA < 200000 & percent.mito < 0.05)




```

## Plots after FILTERING

```{r}


#QC plots
sums <- data.frame(nCountRNA=dataset.fs$nCount_RNA)
p1=ggplot(sums, aes(x=nCountRNA)) + geom_histogram(binwidth=10000)+
  geom_vline(xintercept = mean(sums$nCountRNA), color = "red", size=1)+
  geom_vline(xintercept = median(sums$nCountRNA), color = "blue", size=1)
p1
sums <- data.frame(nb_transcript=dataset.fs$nFeature_RNA)
p2=ggplot(sums, aes(x=nb_transcript)) + geom_histogram(binwidth=100)+
  geom_vline(xintercept = mean(sums$nb_transcript), color = "red", size=1)
p2



# violin plots , features

VlnPlot(object = dataset.fs, features=c("nCount_RNA","nFeature_RNA","percent.mito"),slot = "counts",idents=c("t35-WNT", "t35-BMP","t36-WNT", "t36-BMP","t37-WNT", "t37-BMP"))
VlnPlot(object = dataset.fs, features=c("nCount_RNA","nFeature_RNA","percent.mito"),slot = "counts", group.by = "time")

#VlnPlot(object = dataset.fs, features=c("nCount_RNA","nFeature_RNA","percent.mito"),slot = #"counts",idents=c("t35-WNT", "t35-BMP","t36-WNT", "t36-BMP","t37-WNT", "t37-BMP"))

# scatter plots 

FeatureScatter(object = dataset.fs, feature1 = "nCount_RNA", feature2 = "nFeature_RNA",slot = "counts")

# GAPDH (Glyceraldehyde-3-phosphate dehydrogenase) is commonly considered a housekeeping gene because it is involved in essential cellular functions,
FeatureScatter(object = dataset.fs, feature1 = "nCount_RNA", feature2 = "GAPDH",
               slot = "counts")

```

# ---

-   t36 and t37 are from the same plate .

-   WNT stimulation is seemingly down-regulating genes on average , compared to BMP stimulation .

-   low MT transcript count in t37-wnt

## Normalize and scale

```{r}
## Scale the data to have 80000 transcript (the mean is 85000) in each cell and then, normalizing the data with log1p
#NOTE : adds the data slot 
dataset.fs <- NormalizeData(object = dataset.fs, verbose = FALSE,scale.factor = 80000)

#Next, we apply a linear transformation (‘scaling’) that is a standard pre-processing step prior to dimensional reduction techniques like PCA. The ScaleData() function:
# Shifts the expression of each gene, so that the mean expression across cells is 0
# Scales the expression of each gene, so that the variance across cells is 1
# This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate
dataset.fs <- ScaleData(object = dataset.fs, features = row.names(dataset.fs[['RNA']]$counts))

```

```{r}
library(ggpubr)

## scale with nCount vs  scale withGapdh
for (gene in c("T","EOMES","POU5F1")) {
  data_x <- dataset.fs@assays$RNA$data[gene,]
  data_y <- log1p(max(dataset.fs@assays$RNA$counts[gene,])*dataset.fs@assays$RNA$counts[gene,]/dataset.fs@assays$RNA$counts["GAPDH",])
  c <- dataset.fs$condition
  df <- data.frame(x=data_x,y=data_y,condition=c)
  print(df)
  print(ggplot(data=df, aes(x=x, y=y)) +
    geom_smooth(method="lm") +
    geom_point() +
    stat_regline_equation(label.x=0.1*max(data_x), label.y=0.84*max(data_y)) +
    stat_cor(aes(label=..rr.label..), label.x=0.1*max(data_x), label.y=0.76*max(data_y))+
    xlab(substitute(paste("log(",italic(p), ") (Scale with nCounts)", sep=""), list(p=str_to_title(tolower(gene)))))+
    ylab(substitute(paste("log(",italic(p), ") (Scale with ",italic(q),")", sep=""), list(p=str_to_title(tolower(gene)),q="Gapdh"))))
}

```

```{r}
dataset.fs <- FindVariableFeatures(dataset.fs, selection.method = "vst", nfeatures = 2000)
top20 <- head(VariableFeatures(dataset.fs),20)
top20
plot1 <- VariableFeaturePlot(dataset.fs)
plot2 <- LabelPoints(plot = plot1, points = top20, repel = TRUE)
plot2

```

## PCA :

By default, only the previously determined variable features are used as input, but can be defined using `features` argument if you wish to choose a different subset

```{r}
# Perform linear dimensional reduction
dataset.fs <- RunPCA(dataset.fs, npcs = 30, verbose = FALSE)
PCAPlot(dataset.fs)
PCAPlot(dataset.fs, group.by="condition")
# Examine and visualize PCA results 
# For the first principal components, Seurat outputs a list of genes with the most positive and negative loadings, representing modules of genes that exhibit either correlation (or anti-correlation) across single-cells in the dataset.
print(dataset.fs[["pca"]], dims = 1:3, nfeatures = 3)

# loadings 
# The loadings in PCA refer to the contributions of each gene (or feature) to the principal components (PCs)
as.data.frame(dataset.fs@reductions$pca@feature.loadings)
# embedding are the projections in lower dimensions 
as.data.frame(dataset.fs@reductions$pca@cell.embeddings)


```

```{r}
# List of top genes in PCx
PCx <- 2
top20.neg <- names(sort(dataset.fs@reductions$pca@feature.loadings[,PCx]))[1:20]

# JackStrawpLOT 
# dataset.fs <- JackStraw(dataset.fs,dims = 30)
# dataset.fs <- ScoreJackStraw(dataset.fs, dims = 1:30)
# JackStrawPlot(dataset.fs,dims = 1:30)

# Dimloadings 
VizDimLoadings(dataset.fs, dims = 1:2, reduction = "pca")

# Dim heatmaps 
DimHeatmap(dataset.fs, dims = 1:2, cells = 500, balanced = TRUE)
#Both cells and features are ordered according to their PCA scores. Setting cells to a number #plots the ‘extreme’ cells on both ends of the spectrum, which dramatically speeds plotting for #large datasets.")

# ElbowPlot 
ElbowPlot(dataset.fs)

```

> How to interpret the heatmaps ?

## UMAP

-   As `n_neighbors` increases, UMAP connects more and more neighboring points when constructing the graph representation of the high-dimensional data, which leads to a projection that more accurately reflects the global structure of the data.At very low values, any notion of global structure is almost completely lost (global structure is the 3D mammoth for example [Mammoth plot](https://pair-code.github.io/understanding-umap/) )

-   As the `min_dist` parameter increases, UMAP tends to "spread out" the projected points, leading to decreased clustering of the data and less emphasis on global structure.

    **dims**

    :   Which dimensions to use as input features, used only if `features` is NULL

    **reduction**

    :   Which dimensional reduction (PCA or ICA) to use for the UMAP input. Default is PCA

```{r}
nb_pc <- 15
k <- 10
dataset.fs <- RunUMAP(dataset.fs, reduction = "pca", dims = 1:nb_pc,n.neighbors = k,
                       n.components=2)

DimPlot(dataset.fs, reduction = "umap",dims=c(1,2),group.by = "orig.ident")
DimPlot(dataset.fs, reduction = "umap",dims=c(1,2),group.by = "time")
DimPlot(dataset.fs, reduction = "umap",dims=c(1,2),group.by = "condition")
```

### Find neighbors and cluster .

-   [`FindNeighbors()`](https://satijalab.org/seurat/reference/findneighbors) takes as input the previously defined dimensionality of the dataset (first 10 PCs) : we first construct a KNN graph based on the euclidean distance in PCA space

> -   **KNN (algorithm)** is used for predictive tasks like classification and regression.
>
> -   **KNN graph** is a versatile tool for analyzing relationships between data points, commonly applied in clustering and other unsupervised tasks.

```{r}
## Cluster the cells
resolution = 0.1
k <- 10 # 
dataset.fs <- FindNeighbors(dataset.fs, reduction= "pca", dims = 1:nb_pc, k.param=k)
dataset.fs <- FindClusters(object = dataset.fs, reduction= "pca", resolution = resolution)

DimPlot(dataset.fs, reduction = "umap",label=TRUE)

```

### Use different resolutions to cluster points

```{r}


library(clustree)

## Clusters stability 
for (res in c(1:10)/10){
  dataset.fs <- FindClusters(object = dataset.fs, reduction= "pca", resolution = res)
}

# size=10
# clustree(dataset.fs, prefix = "RNA_snn_res.",node_text_size = size*0.25,edge_width = size*0.25)+
#  theme(legend.key.size = unit(size/12.5, 'cm'),
#        legend.text = element_text(size=size*2.5),
#        legend.title = element_text(size=size*3))


# All resolutions 

DimPlot(dataset.fs,group.by = "RNA_snn_res.0.1",label = TRUE)
DimPlot(dataset.fs,group.by = "RNA_snn_res.0.4",label = TRUE)
DimPlot(dataset.fs,group.by = "RNA_snn_res.0.5",label = TRUE)
DimPlot(dataset.fs,group.by = "RNA_snn_res.0.8",label = TRUE)
DimPlot(dataset.fs,group.by = "RNA_snn_res.1",label = TRUE)
DimPlot(dataset.fs,group.by = "condition")
DimPlot(dataset.fs,group.by = "time")



```

# Feature plots

```{r}
Idents(dataset.fs)<-"RNA_snn_res.0.8"
FeaturePlot(dataset.fs,c("ID1","WNT3","WNT3A","nCount_RNA"),order = TRUE)
FeaturePlot(dataset.fs,c("nCount_RNA"),order = TRUE,split.by = "condition" ) 
FeaturePlot(dataset.fs,c("ID1"),order = TRUE,split.by = "condition" )
FeaturePlot(dataset.fs,c("WNT3","WNT3A"),order = TRUE,split.by = "condition" ) 
FeaturePlot(dataset.fs,c("percent.mito","nFeature_RNA"))


VlnPlot(dataset.fs,"nCount_RNA")
VlnPlot(dataset.fs,c("nCount_RNA","nFeature_RNA")) 
```

# ANNOTATE-ASSIGN CELL TYPES

use Gael annotation - available list

```{r}
library(tibble)
# load the processed data from gael
gael_processed_rd_2<-"/Users/alichemkhi/Desktop/data/2dgas-fs/rds/2dgas_t36_fs.rds"
gael_processed <- readRDS(gael_processed_rd_2)


gael_metadata <- gael_processed@meta.data["celltype"] %>%
  rownames_to_column(var = "ident")
dataset_meta <- dataset.fs@meta.data %>%
  rownames_to_column(var = "ident")

merged_df<-left_join(dataset_meta,gael_metadata,by="ident")
merged_df$celltype<-as.character(merged_df$celltype)
merged_df$celltype[is.na(merged_df$celltype)]<-"unknown"
merged_df$celltype<-as.factor(merged_df$celltype)

merged_df <- merged_df %>% column_to_rownames(var = "ident")
dataset.fs.annotated <- dataset.fs

dataset.fs.annotated@meta.data<-merged_df # overwrite metadata

dim(dataset.fs.annotated@meta.data)
dim(merged_df)



DimPlot(dataset.fs.annotated,group.by = "RNA_snn_res.0.8",label = TRUE)
DimPlot(dataset.fs.annotated,group.by = "celltype",label = TRUE)
FeaturePlot(dataset.fs.annotated,"nCount_RNA")
DimPlot(dataset.fs.annotated,group.by = "condition")
DimPlot(dataset.fs.annotated,group.by = "time")

```

# Remove cells - not annotated by Gael

```{r}

# NOTE : the celltypes were collected from Gael object , unknown labels mean that the cell was not in gael's object. 
# Also this group has very low gene count , see featureplot above 

dataset.annotated.filtered.fs <- subset(dataset.fs.annotated, subset = celltype!="unknown"  )
```

## Add mito and ribo averages

```{r}

pattern_genes <- rownames(dataset.annotated.filtered.fs)[grepl("^mt-", rownames(dataset.annotated.filtered.fs), ignore.case = TRUE)]
pattern_matrix <- GetAssayData(dataset.annotated.filtered.fs, layer = "count")[pattern_genes,]
mean_expression <- colMeans(pattern_matrix)
dataset.annotated.filtered.fs <- AddMetaData(
  object = dataset.annotated.filtered.fs,
  metadata = mean_expression, # Mean expression 
  col.name = "mtMean" )


pattern_genes <- rownames(dataset.annotated.filtered.fs)[grepl("^Rps", rownames(dataset.annotated.filtered.fs), ignore.case = TRUE)]
pattern_matrix <- GetAssayData(dataset.annotated.filtered.fs, layer = "count")[pattern_genes,]
mean_expression <- colMeans(pattern_matrix)
dataset.annotated.filtered.fs <- AddMetaData(
  object = dataset.annotated.filtered.fs,
  metadata = mean_expression, # Mean expression 
  col.name = "rpsMean" )


pattern_genes <- rownames(dataset.annotated.filtered.fs)[grepl("^Rpl", rownames(dataset.annotated.filtered.fs), ignore.case = TRUE)]
pattern_matrix <- GetAssayData(dataset.annotated.filtered.fs, layer = "count")[pattern_genes,]
mean_expression <- colMeans(pattern_matrix)
dataset.annotated.filtered.fs <- AddMetaData(
  object = dataset.annotated.filtered.fs,
  metadata = mean_expression, # Mean expression 
  col.name = "rplMean" )





```

## Save Seurat object

```{r}
dataset.annotated.filtered.fs
saveRDS(dataset.annotated.filtered.fs, file = paste(OUTDIR,"dataset.filtered.rds",sep ="/"))


# Save the metadata as CSV
write.csv(dataset.annotated.filtered.fs@meta.data, file = paste(OUTDIR, "metadata.csv", sep = "/"), row.names = FALSE)

# Save the raw count matrix as CSV
raw_counts<-dataset.annotated.filtered.fs@assays$RNA$counts
write.csv(raw_counts, file = paste(OUTDIR, "raw_counts.csv", sep = "/"), row.names = TRUE)

```

# Merge cell types 

```{r}

dataset.path<- file.path(OUTDIR,"dataset.filtered.rds")
dataset.so     <- readRDS(file =dataset.path)


dataset.so$celltype_grouped <- case_when(
  dataset.so$celltype %in% c("NasMes1", "NasMes2", "NasMes3") ~ "NasMes",
  dataset.so$celltype %in% c("aPS-DE","AxMes") ~ "aPS",
  dataset.so$celltype %in% c("PS1", "PS2") ~ "PS",
  dataset.so$celltype =="Epi" ~ "Epi",
  dataset.so$celltype =="ExeMes" ~ "ExeMes",
  dataset.so$celltype =="PGC" ~ "PGC",
 
)

table(dataset.so@meta.data$celltype)
table(dataset.so@meta.data$celltype_grouped)

saveRDS(dataset.so, file = paste(OUTDIR,"dataset.filtered.grouped.rds",sep ="/"))


```
